servers/src/main/java/tachyon/master/lineage/LineageMaster.java

```diff
@@ -0,0 +1,502 @@
+      requestFilePersistenceFromEntry((RequestFilePersistenceEntry) entry);
+    } else if (entry instanceof DeleteLineageEntry) {
+      deleteLineageFromEntry((DeleteLineageEntry) entry);
+    } else {
+      throw new IOException(ExceptionMessage.UNEXPECETD_JOURNAL_ENTRY.getMessage(entry));
+    }
+  }
+
+  @Override
+  public void start(boolean isLeader) throws IOException {
+    super.start(isLeader);
+    if (isLeader) {
+      mCheckpointExecutionService =
+          getExecutorService().submit(new HeartbeatThread("Checkpoint scheduling service",
+              new CheckpointSchedulingExcecutor(this),
+              mTachyonConf.getInt(Constants.MASTER_LINEAGE_CHECKPOINT_INTERVAL_MS)));
+      mRecomputeExecutionService =
+          getExecutorService().submit(new HeartbeatThread("Recompute service",
+              new RecomputeExecutor(new RecomputePlanner(mLineageStore, mFileSystemMaster),
+                  mFileSystemMaster),
+              mTachyonConf.getInt(Constants.MASTER_LINEAGE_RECOMPUTE_INTERVAL_MS)));
+    }
+  }
+
+  @Override
+  public void stop() throws IOException {
+    super.stop();
+    if (mCheckpointExecutionService != null) {
+      mCheckpointExecutionService.cancel(true);
+    }
+    if (mRecomputeExecutionService != null) {
+      mRecomputeExecutionService.cancel(true);
+    }
+  }
+
+  @Override
+  public synchronized void streamToJournalCheckpoint(JournalOutputStream outputStream)
+      throws IOException {
+    mLineageStore.streamToJournalCheckpoint(outputStream);
+  }
+
+  /**
+   * @return a lineage store view wrapping the contained lineage store
+   */
+  public LineageStoreView getLineageStoreView() {
+    return new LineageStoreView(mLineageStore);
+  }
+
+  /**
+   * Creates a lineage. It creates a new file for each output file.
+   *
+   * @param inputFiles the input files
+   * @param outputFiles the output files
+   * @param job the job
+   * @return the id of the created lineage
+   * @throws InvalidPathException if the path to the input file is invalid
+   * @throws FileAlreadyExistException if the output file already exists
+   * @throws BlockInfoException if fails to create the output file
+   */
+  public synchronized long createLineage(List<TachyonURI> inputFiles, List<TachyonURI> outputFiles,
+      Job job) throws InvalidPathException, FileAlreadyExistException, BlockInfoException {
+    // validate input files exist
+    List<TachyonFile> inputTachyonFiles = Lists.newArrayList();
+    for (TachyonURI inputFile : inputFiles) {
+      long fileId;
+      fileId = mFileSystemMaster.getFileId(inputFile);
+      inputTachyonFiles.add(new TachyonFile(fileId));
+    }
+    // create output files
+    List<LineageFile> outputTachyonFiles = Lists.newArrayList();
+    for (TachyonURI outputFile : outputFiles) {
+      long fileId;
+      // create the file initialized with block size 1 as placeholder
+      fileId = mFileSystemMaster.createFile(outputFile, 1, true);
```

Comment:
if we fail in the middle of this loop, say a file exists, shall we roll back all those created placeholder files?
Otherwise, this `createLineage` operation will throw `FileAlreadyExistException` without cleaning them up, and those 1-byte placeholder files seem totally legitimate since then.